### ISSUE TITLE
Incorrect type-error with interface and throws

### ISSUE THREAD

**Issue by thedrlambda** [microsoft/TypeScript#51102]
# Bug Report

### 🔎 Search Terms

interface, throw

### 🕗 Version & Regression Information

My version is 4.8.2

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about throw and type-error

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMoHsC2ECS5rxLIDeAsAFDLIzroAUAlAFzIDOYUoA5sgD7IBXEABMIMUBGEBuCgF8KCADZxWrNFlz5YiCABEBmTAE9kwTAAdFEbODUZseSNqJlK1Wo2QB6AFTIWIBAAbtDIPl4kFFRUYAAWUOgA7sgARKix6AKKwsgg6GDIAEYoCHCKVsIpMm5UXhFQEGACUCCpVVHI8uRdSip2Go4EOqhNhcgQAB6QIv0OWoR6BsaRbjT0DCvRyA1NLansAoXtbl2yQA)

### 💻 Code

```ts
interface SomeInterface {
  foo(): string | undefined;
}
class SomeInterfaceDummy implements SomeInterface {
  foo() /* : never */ {
    throw "Should not be called";
    // return "";
  }
}
class SomeInterfaceStub extends SomeInterfaceDummy {
  foo() {
    return "stub";
  }
}
```

### 🙁 Actual behavior

There is no way to type-check the code above. The return type is inferred as void (I expected `never`). Adding the out-commented return gives the correct return type but errors with "unreachable code." Putting the return type `: never` explicitly means the inheriting method doesn't compile.

### 🙂 Expected behavior

Body that throws is inferred to have type `never` and `never <: any` so there should be no errors.

**Comment by thedrlambda** [comment]
For now, I am using this ugly workaround:
```ts
  foo(): ReturnType<SomeInterface["foo"]> {
    throw "Should not be called";
  }
```

**Comment by RyanCavanaugh** [comment]
A function consisting only of a `throw` statement is given type `void` for exactly the scenario of a psuedo-abstract base class method. If we did what you propose

 > Body that throws is inferred to have type never and never <: any so there should be no errors.

then `SomeInterfaceStub` would have an error, because `string` is not a subtype of `never`.

**Comment by thedrlambda** [comment]
But surely the function, being inherited from an interface, should preserve the inherited type. I'm not saying you should give the function return type never, I'm suggesting that the body is inferred to have type never, and the function keeps its inherited type, which should type check because never is a subtype of anything.

**Comment by fatcerberus** [comment]
> But surely the function, being inherited from an interface, should preserve the inherited type.

Method argument and return types are not inferred through inheritance.  See https://github.com/microsoft/TypeScript/issues/32082 for that.

**Comment by RyanCavanaugh** [comment]
> and the function keeps its inherited type, which should type check because never is a subtype of anything.

It's not legal to return a value from a function that's supposed to be of typed `never`

**Comment by RyanCavanaugh** [comment]
Ah, right. Of course, that just creates a different problem:
```ts
interface Foo { bar(): string | undefined }
class MyFoo() implements Foo {
  bar() {
    return "hello, world";
  }
}
const m = new MyFoo();
// m.bar() is possibly undefined -- no it isn't??
const s: string = m.bar();
```

**Comment by thedrlambda** [comment]
I don't see the problem in your last comment. `m.bar()` is possibly undefined. Particularly if you use another implementing class than MyFoo. In my original code, I expect:

SomeInterface.foo: string | undefined
SomeInterfaceDummy.foo: string | undefined
SomeInterfaceDummy.foo.body: never (<: string | undefined)
SomeInterfaceStub.foo: string | undefined
SomeInterfaceDummy.foo.body: string (<: string | undefined)

Thus it should all type check.

**Comment by RyanCavanaugh** [comment]
> I don't see the problem in your last comment. m.bar() is possibly undefined

No, it isn't. Because `m` has type `MyFoo`, and `MyFoo.bar` never returns undefined.
